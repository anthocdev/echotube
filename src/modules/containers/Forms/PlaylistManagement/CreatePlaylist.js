import React from "react";
import { Button } from "react-bootstrap";
import { connect } from "react-redux";
import * as PlaylistActions from "../../../../actions/PlaylistActions";

class CreatePlaylist extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      PlaylistName: "",
      PlaylistImageLink: "",
      VerifiedImageLink: "",
      ValidImage: false,
      errorMessage: "",
    };

    this.imageLinkInput = React.createRef();
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  async validForm(playlistData) {
    var isValid = true;
    var error = "";
    if (!playlistData.PlaylistName) {
      isValid = false;
      error = error + "Invalid Playlist Name \n";
    }

    if (!playlistData.PlaylistImageLink) {
      isValid = false;
      error = error + "Invalid Image, Please Verify It. \n";
    }

    await this.setState({
      errorMessage: error,
    });
    return isValid;
  }

  handleChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    this.setState({
      [name]: value,
    });
  }

  async handleSubmit(event) {
    event.preventDefault();
    /* Validation Check */
    /* Request Creation To API */
    var playlistObject = {
      PlaylistName: this.state.PlaylistName,
      PlaylistImageLink: this.state.VerifiedImageLink,
    };

    if (await this.validForm(playlistObject)) {
      this.props.createUserPlaylist(playlistObject);
    } else {
      alert(this.state.errorMessage);
    }
  }

  async dispImage() {
    this.setState({
      ValidImage: await ImageVerify(this.state.PlaylistImageLink),
    });

    if (this.state.ValidImage) {
      this.setState({ VerifiedImageLink: this.state.PlaylistImageLink });
    }
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Playlist Name:
          <input
            name="PlaylistName"
            type="text"
            value={this.state.PlaylistName}
            onChange={this.handleChange}
          />
        </label>

        <label>
          Playlist Image Link:
          <input
            name="PlaylistImageLink"
            type="text"
            value={this.state.PlaylistImageLink}
            onChange={this.handleChange}
          />
          <Button onClick={() => this.dispImage()}>Verify Image</Button>
          {this.state.ValidImage ? (
            <div>
              <img
                src={this.state.VerifiedImageLink}
                height="150px"
                width="150px"
              />
            </div>
          ) : (
            <div>Invalid</div>
          )}
        </label>
        <input type="submit" value="Create Playlist" />
      </form>
    );
  }
}

function ImageVerify(url) {
  return new Promise(function (res, rej) {
    let img = new Image();
    img.src = url;

    img.onload = function () {
      res(true);
    };

    img.onerror = function () {
      res(false);
    };
  });
}

function mapStateToProps(state, props) {
  return {
    userPlaylist: state.userPlaylist,
  };
}

export default connect(mapStateToProps, PlaylistActions)(CreatePlaylist);
