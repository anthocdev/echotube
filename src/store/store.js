import { applyMiddleware, createStore, compose } from "redux";
import { createLogger } from "redux-logger";
import { default as thunk } from "redux-thunk";
import { createPromise } from "redux-promise-middleware";
import axios from "axios";

import reducer from "../reducers";
//Pulling JWT Token from local storage
const jwttoken = localStorage.getItem("JWT_TOKEN");
//Setting JWT Token as common header for axios
axios.defaults.headers.common["Authorization"] = jwttoken;
//Create store with middleware included
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose; //Adding redux dev tool compatibility
const middleware = applyMiddleware(createPromise(), thunk, createLogger()); //Adding thunk and console logging for testing as middleware

export default createStore(
  reducer,
  {
    //Initializing auth state if JWT token is available
    auth: {
      jwtToken: jwttoken,
      isAuth: jwttoken ? true : false
    }
  },
  composeEnhancers(middleware)
);
