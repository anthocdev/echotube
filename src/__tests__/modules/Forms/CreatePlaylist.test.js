import React from "react";
import CreatePlaylist from "../../../modules/containers/Forms/PlaylistManagement/CreatePlaylist";
import { Provider } from "react-redux";
import store from "../../../store/store";
import { create, act } from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

describe("Playlist Creation Module Test", () => {
  let root;

  it("Renders Without Crashing", async () => {
    act(() => {
      root = create(
        <Provider store={store}>
          <CreatePlaylist />
        </Provider>
      );
    });
  });

  test("Renders Matches Snapshot", async () => {
    act(() => {
      root = create(
        <BrowserRouter>
          <Provider store={store}>
            <CreatePlaylist />
          </Provider>
        </BrowserRouter>
      );
    });
    expect(root).toMatchSnapshot();
    console.log(root.toJSON());
  });
});
