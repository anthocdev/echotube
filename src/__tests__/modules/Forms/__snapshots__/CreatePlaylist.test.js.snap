// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist Creation Module Test Renders Matches Snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    Playlist Name:
    <input
      name="PlaylistName"
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <label>
    Playlist Image Link:
    <input
      name="PlaylistImageLink"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Verify Image
    </button>
    <div>
      Invalid
    </div>
  </label>
  <input
    type="submit"
    value="Create Playlist"
  />
</form>
`;
